samples/file-encryptor/enclave/openssl_src/keys.cpp

```diff
@@ -0,0 +1,438 @@
+
+    TRACE_ENCLAVE("Done with prepare_encryption_header successfully.");
+    ret = 0;
+exit:
+    return ret;
+}
+
+// Parse an input header for validating the password and getting the encryption
+// key in preparation for decryption/encryption operations
+//  1)Check password by comparing their digests
+//  2)reproduce a password key from the password
+//  3)decrypt the encryption key with a password key
+int ecall_dispatcher::parse_encryption_header(
+    encryption_header_t* header,
+    string password)
+{
+    int ret = 1;
+    if (header == NULL)
+    {
+        TRACE_ENCLAVE("parse_encryption_header() failed with a null argument"
+                      " for encryption_header_t*");
+        goto exit;
+    }
+
+    unsigned char digest[HASH_VALUE_SIZE_IN_BYTES];
+    unsigned char password_key[ENCRYPTION_KEY_SIZE_IN_BYTES];
+    unsigned char salt[SALT_SIZE_IN_BYTES];
+
+    // check password by comparing their digests
+    ret =
+        Sha256((const uint8_t*)m_password.c_str(), m_password.length(), digest);
+    if (ret)
+    {
+        TRACE_ENCLAVE("Sha256 failed with %d", ret);
+        goto exit;
+    }
+
+    if (memcmp(header->digest, digest, HASH_VALUE_SIZE_IN_BYTES) != 0)
+    {
+        TRACE_ENCLAVE("incorrect password");
+        ret = 1;
+        goto exit;
+    }
+
+    memcpy(salt, header->salt, SALT_SIZE_IN_BYTES);
+
+    // derive a key from the password using PBDKF2
+    ret = generate_password_key(
+        password.c_str(), salt, password_key, ENCRYPTION_KEY_SIZE_IN_BYTES);
+    if (ret != 0)
+    {
+        TRACE_ENCLAVE("generate_password_key failed with %d", ret);
+        goto exit;
+    }
+
+    // decrypt the "encrypted encryption key" using the password key
+    ret = cipher_encryption_key(
+        DECRYPT_OPERATION,
+        header->encrypted_key,
+        ENCRYPTION_KEY_SIZE_IN_BYTES,
+        password_key,
+        salt, // iv for encryption, decryption. In this sample we use
+              // the salt in encryption header as iv.
+        (unsigned char*)m_encryption_key,
+        ENCRYPTION_KEY_SIZE_IN_BYTES);
+    if (ret != 0)
+    {
+        TRACE_ENCLAVE("Enclave: m_encryption_key");
+        for (unsigned int i = 0; i < ENCRYPTION_KEY_SIZE_IN_BYTES; i++)
+            TRACE_ENCLAVE(
+                "m_encryption_key[%d] =0x%02x", i, m_encryption_key[i]);
+        goto exit;
+    }
+
+    ret = 0;
+exit:
+    return ret;
+}
+
+int ecall_dispatcher::process_encryption_header(encryption_header_t* header)
+{
+    int ret = 1;
+
+    if (header == NULL)
+    {
+        TRACE_ENCLAVE("process_encryption_header() failed with a null argument"
+                      " for encryption_header_t*");
+        goto exit;
+    }
+
+    if (m_encrypt)
+    {
+        // allocate host memory for the header and it will be returned back to
+        // the host
+        m_header =
+            (encryption_header_t*)oe_host_malloc(sizeof(encryption_header_t));
```

Comment:
oe_host_malloc should never be used in samples, as it's not cross-platform.  Instead use ecalls or calls to return values to the host.

Reply:
I have adopted the same API design as the original version here. Can I address oe_host_malloc deprecation in a different PR?

https://github.com/openenclave/openenclave/issues/1956

Reply:
Yes fine to do in a different PR